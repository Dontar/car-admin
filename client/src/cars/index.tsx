import {
    Datagrid, DateField, DateInput, Edit, EditProps, Filter, FilterProps, List, ListProps, ReferenceField,
    ReferenceInput,
    SelectInput,
    Show, ShowProps, SimpleForm, SimpleShowLayout, TextField, TextInput
} from 'react-admin';

const CarFilter = (props: Partial<FilterProps>) => (
    <Filter {...props}>
        <TextInput label="Car number" source="dkn" alwaysOn />
    </Filter>
);

export const CarList = (props: ListProps) => (
    <List {...props} filters={<CarFilter />}>
        <Datagrid rowClick="show" >
            {/* <TextField source="id" /> */}
            <TextField source="rama" />
            <TextField source="dkn" />
            <TextField source="mark_name" />
            <TextField source="model_name" />
            <DateField source="produce_year" />
            <ReferenceField source="company_id" reference="companies" link="show">
                <TextField source="client_name" display='block' noWrap style={{ textOverflow: 'ellipsis', maxWidth: '5cm' }} />
            </ReferenceField>
            <ReferenceField source="person_id" reference="people" link="show">
                <TextField source="client_name" display='block' noWrap style={{ textOverflow: 'ellipsis', maxWidth: '5cm' }} />
            </ReferenceField>
        </Datagrid>
    </List>
);

export const CarShow = (props: ShowProps) => (
    <Show {...props}>
        <SimpleShowLayout>
            <TextField source="id" />
            <TextField source="rama" />
            <TextField source="dkn" />
            <TextField source="mark_name" />
            <TextField source="model_name" />
            <DateField source="produce_year" />
            <ReferenceField source="company_id" reference="companies">
                <TextField source="id" />
            </ReferenceField>
            <ReferenceField source="person_id" reference="people">
                <TextField source="id" />
            </ReferenceField>
        </SimpleShowLayout>
    </Show>
);

export const CarEdit = (props: EditProps) => (
    <Edit {...props}>
        <SimpleForm>
            <TextInput source="id" />
            <TextInput source="rama" />
            <TextInput source="dkn" />
            <TextInput source="mark_name" />
            <TextInput source="model_name" />
            <DateInput source="produce_year" />
            <ReferenceInput source="company_id" reference="companies">
                <SelectInput optionText="id" />
            </ReferenceInput>
            <ReferenceInput source="person_id" reference="people">
                <SelectInput optionText="id" />
            </ReferenceInput>
        </SimpleForm>
    </Edit>
);