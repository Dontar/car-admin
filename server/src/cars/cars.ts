import { getDB } from "../connection";
import rawSql, { in as $in, like, select } from 'sql-bricks-sqlite';
import { GetOneParams, GetOneResult, GetParams, GetResult, ICars } from "../share/models";
import { getSqlCount, isOne, prepareParams } from "../share/utils";

export async function getCars(params: GetOneParams): Promise<GetOneResult<ICars>>;
export async function getCars(params: Partial<GetParams>): Promise<GetResult<ICars>>;
export async function getCars(params: Partial<GetParams> | GetOneParams): Promise<GetResult<ICars> | GetOneResult<ICars>> {
    const db = await getDB();

    let sql = select(
        'v.id as id',
        'upper(v.rama) as rama',
        'v.dkn as dkn',
        'vm.name as mark_name',
        'vmd.name as model_name',
        'v.produce_year as produce_year',
        'max(vc.client_id) as company_id',
        'max(vc.client_id) as person_id'
    ).from('vehicles v')
        .join('vehicle_mark vm', { 'vm.ID': 'v.MARK_ID' })
        .join('vehicle_model vmd', { 'vmd.ID': 'v.MODEL_ID' })
        .leftJoin('vc', { 'vc.VEHICLE_ID': 'v.ID' })
        .groupBy('v.ID');

    let total: number | undefined = 1376287;
    if (isOne(params)) {
        sql.where({ 'v.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('v.id', ...params.ids));
        } else {
            if (params.filter) {
                const { dkn, ...filter } = params.filter;
                sql.where(like('dkn', rawSql(`'${dkn}%'`)));
                sql.where(filter);
                total = await getSqlCount(sql, db);
            }
            sql = await prepareParams(params, sql);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<ICars>(text, values);
        return {
            data: row!
        };
    } else {
        const rows = await db.all<ICars[]>(text, values);
        return {
            data: rows,
            total
        };
    }
}
