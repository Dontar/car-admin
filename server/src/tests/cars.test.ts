import { config } from 'dotenv';
import { getCars } from "../cars/cars";
config();

describe('Test car data retrieval', () => {

    it('should get a car from db', async () => {
        const data = await getCars({
            id: 4601
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 cars from db', async () => {
        const data = await getCars({
            ids: [4601, 4602]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get cars from db with pagination', async () => {
        const data = await getCars({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get cars from db with sorting', async () => {
        const data = await getCars({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get cars from db with filter', async () => {
        const data = await getCars({
            filter: { dkn: 'СА' },
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

});
