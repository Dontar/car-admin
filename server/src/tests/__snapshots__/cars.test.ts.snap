// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test car data retrieval should get 2 cars from db 1`] = `
Object {
  "data": Array [
    Object {
      "company_id": null,
      "dkn": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "person_id": null,
      "produce_year": "1992",
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "company_id": null,
      "dkn": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "person_id": null,
      "produce_year": "2008",
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;

exports[`Test car data retrieval should get a car from db 1`] = `
Object {
  "data": Object {
    "company_id": null,
    "dkn": "CA4918KM",
    "id": 4601,
    "mark_name": "Volkswagen",
    "model_name": "Passat",
    "person_id": null,
    "produce_year": "1992",
    "rama": "YGHJHKLLOLIHGHG",
  },
}
`;

exports[`Test car data retrieval should get cars from db with filter 1`] = `
Object {
  "data": Array [
    Object {
      "company_id": 1753520,
      "dkn": "СА5375ХМ",
      "id": 1490009,
      "mark_name": "Volkswagen",
      "model_name": "Golf",
      "person_id": 1753520,
      "produce_year": "1994",
      "rama": "WVWZZZ1HZRW464701",
    },
    Object {
      "company_id": 1556492,
      "dkn": "СА8652ХВ",
      "id": 2134851,
      "mark_name": "Volkswagen",
      "model_name": "Golf",
      "person_id": 1556492,
      "produce_year": "2002",
      "rama": "WVWZZZ1JZ1B173308",
    },
    Object {
      "company_id": 1468442,
      "dkn": "СА2525ВК",
      "id": 2143856,
      "mark_name": "Toyota",
      "model_name": "RAV-4",
      "person_id": 1468442,
      "produce_year": "2004",
      "rama": "JTEHG20V306040443",
    },
    Object {
      "company_id": 1552797,
      "dkn": "СА8346ХВ",
      "id": 2168514,
      "mark_name": "Audi",
      "model_name": "A8",
      "person_id": 1552797,
      "produce_year": "2004",
      "rama": "WAUZZZ4E94N020180",
    },
    Object {
      "company_id": 112645,
      "dkn": "СА8393ХВ",
      "id": 2169761,
      "mark_name": "Audi",
      "model_name": "A6",
      "person_id": 112645,
      "produce_year": "2000",
      "rama": "WAUZZZ4BZYN098915",
    },
    Object {
      "company_id": 264536,
      "dkn": "СА1309А",
      "id": 2189941,
      "mark_name": "0",
      "model_name": "Друг ...",
      "person_id": 264536,
      "produce_year": "2001",
      "rama": "JYARP041000004589",
    },
    Object {
      "company_id": 1582154,
      "dkn": "СА0218ХН",
      "id": 2312244,
      "mark_name": "Volkswagen",
      "model_name": "Caddy",
      "person_id": 1582154,
      "produce_year": "2007",
      "rama": "WV2ZZZ2KZ8X059506",
    },
    Object {
      "company_id": 92082,
      "dkn": "СА7568ХН",
      "id": 2357383,
      "mark_name": "Citroen",
      "model_name": "Друг ...",
      "person_id": 92082,
      "produce_year": "2005",
      "rama": "VF7LC9HXC74172495",
    },
    Object {
      "company_id": 1815922,
      "dkn": "СА1453ХР",
      "id": 2378536,
      "mark_name": "BMW",
      "model_name": "Друг ...",
      "person_id": 1815922,
      "produce_year": "1991",
      "rama": "WBACA31030EG37316",
    },
    Object {
      "company_id": 1518384,
      "dkn": "СА5376ХР",
      "id": 2401528,
      "mark_name": "Seat",
      "model_name": "Ibiza",
      "person_id": 1518384,
      "produce_year": "1995",
      "rama": "VSSZZZ6KZZR737515",
    },
  ],
  "total": 18,
}
`;

exports[`Test car data retrieval should get cars from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "company_id": null,
      "dkn": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "person_id": null,
      "produce_year": "1992",
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "company_id": null,
      "dkn": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "person_id": null,
      "produce_year": "2008",
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;

exports[`Test car data retrieval should get cars from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "company_id": null,
      "dkn": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "person_id": null,
      "produce_year": "1992",
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "company_id": null,
      "dkn": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "person_id": null,
      "produce_year": "2008",
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;
